infrastructures/infrastructure-openstack/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/OpenstackInfrastructure.java

```diff
@@ -325,4 +357,241 @@ protected void unregisterNodeAndRemoveInstanceIfNeeded(final String instanceTag,
             return null;
         });
     }
+
+    private void internalAcquireNodes(int numberOfNodes, Map<String, ?> dynamicPolicyParameters) {
+
+        nbOfNodes = numberOfNodes;
+
+        if (nbOfNodes > 0 && checkDynamicPolicyParameters(nbOfNodes, dynamicPolicyParameters)) {
+
+            // Create Openstack infrastructure (if it does not exist) and initialize its persistent variables
+            initializeOpenstackInfrastructure();
+
+            // Determine the number of instances to deploy and check it
+            if (!checkOpenstackInstancesToDeploy()) {
+                return;
+            }
+
+            int nbOfDeployedNodes = 0;
+
+            Map<String, Set<String>> instancesAndNodesToDeploy = Maps.newHashMap();
+
+            for (int i = 0; i < instancesToDeploy; i++) {
+
+                // Determine the numbe rof nodes to start in the current instance
+                int nodesInCurrentInstance = (existingNodes + nbOfDeployedNodes +
+                                              numberOfNodesPerInstance <= maxNodes) ? numberOfNodesPerInstance
+                                                                                    : nbOfNodes - nbOfDeployedNodes;
+                // Determine the instance tag
+                int currentIndex = getIndexAndIncrementWithLockAndPersist();
+                String instanceTag = getInfrastructureId() + "_" + currentIndex;
+                logger.info("Deploying Openstack instance with tag " + instanceTag + " and the number of nodes " +
+                            nodesInCurrentInstance);
+
+                // Build nodes'start scripts and deploy instance
+                List<String> scripts = createScripts(instanceTag, instanceTag, nodesInCurrentInstance);
+                createOpenstackInstance(instanceTag, scripts);
+
+                // Declare deploying nodes
+                Set<String> deployedNodes = declareNodesAsDeploying(nodesInCurrentInstance, instanceTag);
```

Comment:
you might want to store deploying nodes in persistent variables in case of recovery

Reply:
The set 'deployedNodes' is used to save deploying nodes temporary, in order to remove them if the deployment fails. So, no need to save them in persistent variables.